lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,375 @@
+
+      // For any other ops, create input ports for any captured operands.
+      for (auto operand : op.getOperands())
+        if (operand.getParentBlock() != body)
+          createInputPort(operand);
+    }
+
+    // Create the new module.  This grabs a lock to modify the circuit.
+    SmallString<32> newModuleName(moduleOp.getName());
+    newModuleName.append("_");
+    newModuleName.append(groupName);
+    FModuleOp newModule =
+        buildNewModule(builder, group.getLoc(), newModuleName, ports);
+    SymbolTable::setSymbolVisibility(newModule,
+                                     SymbolTable::Visibility::Private);
+    newModule.getBody().takeBody(group.getRegion());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "      New Module: " << newModuleName << "\n";
+      llvm::dbgs() << "        ports:\n";
+      for (size_t i = 0, e = ports.size(); i != e; ++i) {
+        auto port = ports[i];
+        auto value = connectValues[i];
+        llvm::dbgs() << "          - name: " << port.getName() << "\n"
+                     << "            type: " << port.type << "\n"
+                     << "            direction: " << port.direction << "\n"
+                     << "            value: " << value << "\n";
+      }
+    });
+
+    // Replace the original group with an instance.  Hook up the instance.
+    builder.setInsertionPointAfter(group);
+    auto moduleName = newModule.getModuleName();
+    auto instanceOp = builder.create<InstanceOp>(
+        group.getLoc(), /*moduleName=*/newModule,
+        /*name=*/
+        (Twine((char)tolower(moduleName[0])) + moduleName.drop_front()).str(),
+        NameKindEnum::DroppableName,
+        /*annotations=*/ArrayRef<Attribute>{},
+        /*portAnnotations=*/ArrayRef<Attribute>{}, /*lowerToBind=*/true);
+    instanceOp->setAttr("output_file",
+                        hw::OutputFileAttr::getFromFilename(
+                            builder.getContext(),
+                            "groups_" + circuitName + "_" + groupName + ".sv",
+                            /*excludeFromFileList=*/true));
+    createdOps.insert(instanceOp);
+
+    // Connect instance ports to values.
+    assert(ports.size() == connectValues.size() &&
+           "the number of instance ports and values to connect to them must be "
+           "equal");
+    for (unsigned portNum = 0, e = newModule.getNumPorts(); portNum < e;
+         ++portNum) {
+      OpBuilder::InsertionGuard guard(builder);
+      builder.setInsertionPointAfterValue(instanceOp.getResult(portNum));
+      if (instanceOp.getPortDirection(portNum) == Direction::In)
+        builder.create<StrictConnectOp>(builder.getUnknownLoc(),
+                                        instanceOp.getResult(portNum),
+                                        connectValues[portNum]);
+      else
+        builder.create<StrictConnectOp>(
+            builder.getUnknownLoc(), connectValues[portNum],
+            builder.create<RefResolveOp>(builder.getUnknownLoc(),
```

Comment:
Maybe use the instance location (would suggest the port locations if those were tracked for instances)?

Reply:
I audited these and should have everything correct. With some updates to missing emitter locations for modules and ports, this can now be shown. 

Input FIRRTL:
```
FIRRTL version 3.1.0
circuit Foo: %[[
  {"class":"firrtl.transforms.DontTouchAnnotation", "target":"~Foo|Foo>A_a"},
  {"class":"firrtl.transforms.DontTouchAnnotation", "target":"~Foo|Foo>B_a"}
]]
  declgroup A, bind:
    declgroup B, bind:
    declgroup C, bind:

  module Foo:
    input a: UInt<1> @[a.scala 1:1]
    input b: UInt<1> @[b.scala 2:2]

    group A:          @[group_A.scala 4:4]
      node A_a = a    @[node_A_a 5:5]
      group B:        @[group_A_B.scala 6:6]
        node B_a = a  @[node_B_a 7:7]
```

FIRRTL after this pass (manually stripping sv.verbatim using `firtool groups/Foo.fir -ir-fir -mlir-print-debuginfo | sed 's/^ *sv.verbatim.*//' | circt-translate -export-firrtl`:
```
FIRRTL version 3.1.0
circuit Foo :
  module Foo_A_B : @[group_A_B.scala 6:6]
    input _a : UInt<1> @[node_B_a 7:7]

    node B_a = _a @[node_B_a 7:7]

  module Foo_A : @[group_A.scala 4:4]
    input _a : UInt<1> @[node_A_a 5:5]
    output _foo_A_B__a : Probe<UInt<1>> @[node_B_a 7:7]

    define _foo_A_B__a = probe(_a) @[node_B_a 7:7]
    node A_a = _a @[node_A_a 5:5]

  module Foo : @[groups/Foo.fir 10:10]
    input a : UInt<1> @[a.scala 1:1]
    input b : UInt<1> @[b.scala 2:2]

    inst foo_A_B of Foo_A_B @[group_A_B.scala 6:6]
    inst foo_A of Foo_A @[group_A.scala 4:4]
    connect foo_A_B._a, read(foo_A._foo_A_B__a) @[node_B_a 7:7]
    connect foo_A._a, a @[node_A_a 5:5]
```

Does this look correct to you?

Reply:
My idea here is to have all the ports, connects, and ref ops that are created take the source locator from the original capture. I.e., all of these should be either `node_A_a` or `node_B_a`.

Reply:
Awesome, thank you for this care and for chasing this down! :rocket: